authorization_test.go

```diff
@@ -0,0 +1,119 @@
+package certauth_test
+
+import (
+	"fmt"
+	"testing"
+
+	. "github.com/pantheon-systems/go-certauth"
+)
+
+func TestAuthValidateOU(t *testing.T) {
+	tests := []struct {
+		AllowedOUs []string
+		ActualOUs  []string
+		IsAllowed  bool
+	}{
+		{[]string{}, []string{"endpoint"}, true},
+		{[]string{"endpoint"}, []string{"endpoint"}, true},
+		{[]string{"endpoint"}, []string{"site"}, false},
+		{[]string{"endpoint"}, []string{""}, false},
+		{[]string{"endpoint", "titan"}, []string{"site"}, false},
+		{[]string{"endpoint", "titan"}, []string{"titan"}, true},
+	}
+
+	for _, tc := range tests {
+		check := AllowSpecificOUandCNs{OUs: tc.AllowedOUs, CNs: nil}
+		err := check.IsAuthorized(tc.ActualOUs, "")
+
+		if err != nil && tc.IsAllowed {
+			t.Errorf("Expected AllowedOUs (%v) and ActualOUs (%v) to pass validation, but it failed: err: %s", tc.AllowedOUs, tc.ActualOUs, err)
+		}
+
+		if err == nil && !tc.IsAllowed {
+			t.Errorf("Expected AllowedOUs (%v) and ActualOUs (%v) to fail validation, but it passed.", tc.AllowedOUs, tc.ActualOUs)
+		}
+	}
+}
+
+func TestAuthValidateCN(t *testing.T) {
+	tests := []struct {
+		AllowedCNs []string
+		ActualCN   string
+		IsAllowed  bool
+	}{
+		{[]string{}, "site1", true},
+		{[]string{"site1"}, "site1", true},
+		{[]string{"site1"}, "site", false},
+		{[]string{"site1"}, "", false},
+		{[]string{"site1", "site2"}, "site1", true},
+		{[]string{"site1", "site2"}, "site2", true},
+		{[]string{"site1", "site2"}, "site3", false},
+	}
+
+	for _, tc := range tests {
+		check := AllowSpecificOUandCNs{OUs: nil, CNs: tc.AllowedCNs}
+		err := check.IsAuthorized([]string{""}, tc.ActualCN)
+
+		if err != nil && tc.IsAllowed {
+			t.Errorf("Expected AllowedCNs (%v) and ActualCN (%v) to pass validation, but it failed: err: %s", tc.AllowedCNs, tc.ActualCN, err)
+		}
+
+		if err == nil && !tc.IsAllowed {
+			t.Errorf("Expected AllowedCNs (%v) and ActualCN (%v) to fail validation, but it passed.", tc.AllowedCNs, tc.ActualCN)
+		}
+	}
+}
+
+func TestParsesCNProperly(t *testing.T) {
+	tests := []struct {
+		InputCN      string
+		ExpectedSite string
+		ExpectedEnv  string
+		ExpectedErr  error
+	}{
+		{
+			"dev.de7ad059-19dd-4e45-9095-ef7507d8195b.pantheon.com",
+			"de7ad059-19dd-4e45-9095-ef7507d8195b",
+			"dev",
+			nil,
+		},
+		{
+			"foobar.9eb5ca28-09be-45b9-8068-7ed6af62fcad.some.website.edu",
+			"9eb5ca28-09be-45b9-8068-7ed6af62fcad",
+			"foobar",
+			nil,
+		},
+		{
+			"I_AM_A_CN",
+			"",
+			"",
+			fmt.Errorf(`Unexpected CN format: "I_AM_A_CN"`),
+		},
+		{
+			"dev.myspecialsite1.pantheon.com",
+			"",
+			"",
+			fmt.Errorf(`Site ID (from CN) is not a valid UUID: "myspecialsite1"`),
+		},
+	}
+
+	for _, tc := range tests {
+		actualSite, actualEnv, actualErr := ParseSiteEnvFromCN(tc.InputCN)
+
+		switch {
+		case actualErr == nil && tc.ExpectedErr != nil:
+			fallthrough
+		case actualErr != nil && tc.ExpectedErr == nil:
+			fallthrough
+		case actualErr != nil && tc.ExpectedErr != nil &&
```

Comment:
oh, sorry and one more case would be needed:
```
case actualErr == nil && tc.ExpectedErr == nil:
			break
```

Reply:
meh

Reply:
Probably this would be more clear if I wrote an `ErrIsEqual` helper that did all this junk to make sure they're both not null and then checking `.Error()` to ensure they match. But all this is required, simply checking that they're both nil isn't sufficient.

Reply:
Yea. Don't know if need a helper for one-time usage. Let's keep it as it is.